### PR TITLE
feat: BNP12

### LINKED ISSUE #1004: BNP12 (ec points altbn128 extendend on BNF12)
# What 
BNP12 is necessary to implement the ec_pairing.
Those class/struct need to be implemented in cairo 

# How

Create the struct representing a point on the altbn128 elliptic curve extended on `BNF12`:
```
struct BNP12 {
   x: BNF12,
   y: BNF12,
}

// Used in is_on_curve and double
func A() {
 return <BNF12 Zero element>
}

// Used in is_on_curve
func B() {
   return <BNF12.from_int(3)>
}
```

### PR DESCRIPTION
**Pull request by obatirou**
close #1004

### CODE REVIEW

File: cairo/ethereum/crypto/alt_bn128.cairo [hunk 1]
```diff
@@ -352,3 +352,59 @@ func BNF12_I_PLUS_9() -> BNF12 {
...
+struct BNP12Struct {
+    x: BNF12,
+    y: BNF12,
+}
+
+struct BNP12 {
+    value: BNP12Struct*,
+}
+
+func A() -> BNF12 {
```
Comment (zmalatrax) [inline cairo/ethereum/crypto/alt_bn128.cairo:387]: Add a comment to explain what it is (weierstrass curve coeffs ?)

File: cairo/ethereum/crypto/alt_bn128.cairo [hunk 2]
```diff
@@ -352,3 +352,59 @@ func BNF12_I_PLUS_9() -> BNF12 {
...
+    let bnf12_zero = BNF12_ZERO();
+    return bnf12_zero;
+}
+
+func B() -> BNF12 {
```
Comment (zmalatrax) [inline cairo/ethereum/crypto/alt_bn128.cairo:392]: add a similar comment as `func A()`

File: cairo/tests/utils/args_gen.py [hunk 3]
```diff
@@ -989,6 +990,22 @@ def _gen_arg(
         segments.load_data(base, coeffs)
         return base
 
+    if arg_type is BNP12:
+        if arg is None:
+            return 0
```
Comment (zmalatrax) [inline cairo/tests/utils/args_gen.py:995]: Why `return 0`, and not `return None` ?

File: cairo/tests/utils/strategies.py [hunk 4]
```diff
@@ -526,6 +526,63 @@ def bounded_u256_strategy(min_value: int = 0, max_value: int = 2**256 - 1):
...
+    For alt_bn128, we can use the formula: sqrt(a) = a^((p+1)/4) mod p
+    Use Euler's criterion to check if a is a quadratic residue.
+    Returns None if a is not one.
+    """
+    if pow(a, (p - 1) // 2, p) != 1:
+        return None
+    return pow(a, (p + 1) // 4, p)
+
+
+def generate_valid_point(x_value):
```
Comment (zmalatrax) [inline cairo/tests/utils/strategies.py:555]: The function is specific to `BNP12`, on the alt_bn128 curve ; I think it should be specified in the function signature (+ a small docstring)

File: cairo/tests/utils/strategies.py [hunk 5]
```diff
@@ -526,6 +526,63 @@ def bounded_u256_strategy(min_value: int = 0, max_value: int = 2**256 - 1):
...
+
+    return BNP12(BNF12(x_coords), BNF12(y_coords))
+
+
+# Strategy for BNP12 points on the curve
+bnp12_strategy = st.one_of(
+    st.just(bnp12_infinity),
+    st.just(bnp12_g1),
+    # Generate points with correctly computed y-coordinates
+    st.integers(min_value=1, max_value=1000)
```
Comment (zmalatrax) [inline cairo/tests/utils/strategies.py:580]: Why x is limited to $`[1..1000]`$ ?

File: cairo/tests/utils/args_gen.py (hunk 3 above)
Comment (obatirou) [reply cairo/tests/utils/args_gen.py:995]: There is no need for this indeed. 
arg cannot be None here

File: cairo/tests/utils/strategies.py (hunk 5 above)
Comment (obatirou) [reply cairo/tests/utils/strategies.py:580]: only for testing efficiency. We could modify the strategy to include more edge cases indeed.

Comment (zmalatrax) [reply cairo/tests/utils/strategies.py:580]: Okok, `BNP12` is based on `BNF12` whose strategy already cover the whole spectrum $`[0..p-1]`$
If testing perf are impacted, it could be enough for `BNP12` to rely on `BNF12` for testing edge cases 🤔

**Review by zmalatrax — APPROVED** [review]

File: cairo/tests/utils/strategies.py (hunk 5 above)
Comment (obatirou) [reply cairo/tests/utils/strategies.py]: In fact, refactored to take into account the [0, p-1]. Testing efficiency is not necessary for now (no impact here)

**Review by zmalatrax — APPROVED** [review]
